### PR TITLE
[backport/1.23] wasm: update Wasmtime to v0.39.1. (#22490)

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Piotr Sikora <piotrsikora@google.com>

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
cc @kfaseela @PiotrSikora @mpwarres

**Comment by phlax** [comment]
i think this will at least require a backport of `proxy-wasm-cpp-host` but probably more to pass CI

**Comment by phlax** [comment]
@mattklein123 this one im a bit less confident about ive backported/updated the dependency to resolve issues flagged by the CVE scanner, but ive had to pick very specific commits to pass CI and there are other updates around the same time

@ingwonsong would you be able to look this over?

**Review by mattklein123 — DISMISSED** [review]

**Comment by ingwonsong** [comment]
@phlax Looks good to me. Thanks!

**Comment by phlax** [comment]
seems like its hit further issues in CI - any ideas which commits i need to pick to make it work ?

**Comment by phlax** [comment]
@ingwonsong seems like its actually failing in CI 

```console
error[E0599]: no variant or associated item named `SharedMemory` found for enum `Extern` in the current scope
  --> external/com_github_wasmtime/crates/c-api/src/extern.rs:23:17
   |
23 |         Extern::SharedMemory(_) => todo!(),
   |                 ^^^^^^^^^^^^ variant or associated item not found in `Extern`

error[E0599]: no variant or associated item named `SharedMemory` found for enum `Extern` in the current scope
```

File: bazel/repository_locations.bzl [hunk 1]
```diff
@@ -1149,8 +1149,8 @@ REPOSITORY_LOCATIONS_SPEC = dict(
         project_name = "WebAssembly for Proxies (C++ host implementation)",
         project_desc = "WebAssembly for Proxies (C++ host implementation)",
         project_url = "https://github.com/proxy-wasm/proxy-wasm-cpp-host",
-        version = "2cf68d6c00b4141ecf2afb5e5d51ebd21a27ed43",
-        sha256 = "6314d3e36e8aa9c4e792772b66f773fcfd58affcd4fd437c6d3dc540165e7fc6",
+        version = "4ddbed3c8c8c1aa76db8157e22ae3be676e76ac1",
```
Comment (PiotrSikora) [inline bazel/repository_locations.bzl:1152]: Wastime was updated to v0.39.1 in https://github.com/proxy-wasm/proxy-wasm-cpp-host/commit/9387a5b77418e41bbf2b0909c68c14db42a1e743.

Comment (phlax) [reply bazel/repository_locations.bzl]: i added the commit from #22469 as it was getting `wasm_config` errors in the build without this

ive kept that commit but reverted the `proxy-wasm-cpp-host` to the one from the 0.39.1 update - lets see ..

Comment (phlax) [reply bazel/repository_locations.bzl]: great seems to have worked - thanks @PiotrSikora

**Comment by kfaseela** [comment]
@phlax thanks!

**Pull request merged**